### DISCUSSION TITLE
Accuracy adjustment (manual, automatic or temperature compensated)

### DISCUSSION THREAD

**Discussion by pebblewidower** [joeycastillo/Sensor-Watch#74] [Ideas]
First of all, sorry if this is not the appropriate place to post this. I am a criminal lawyer and I don't know any programming beyond basic HTML (:

Second, for those unfamiliar with what I am talking about, please take a look at [this post](https://www.watchuseek.com/threads/thermocompensation-methods-and-movements.2087/post-10233), especially the part on the digital count adjustment method (it comes right after Figure 9, the picture of a Rolex, but the whole post is worth reading).

I have already posed this question in more basic terms during fundraising, but I wanted to discuss this further. Would it be possible for the real-time clock to get readings from the thermistor and do this sort of digital count adjustment? This would be displayed, naturally, on Basic Clock. According to @josecastillo, the RTC can be adjusted in steps of 1ppm, which equals to ~31 seconds per year. That is not a very fine adjustment (a $3200 Grand Seiko can be adjusted in steps of 6-8spy) but way better than just 'good enough'.

Even if it is not possible or hard to implement temperature compensation (since there would have to be a rough temperature/accuracy drift table), there is another way to increase the accuracy of the Sensor Watch. A Belarusian watch manufacturer called Technochas has this very rudimentary method of adjusting digital quartz watches. Their classic 52-55 model line has an option, in settings, of [slowing or speeding up the clock](https://youtu.be/U2R650QQIW0?t=138) in steps of 1/10th of a second per day. This equals to roughly 1ppm, which is possible with Sensor Watch. Basically you set the watch to a reference clock, check the deviation after a few days, calculate how much it lost or gained in average per day, and input this in the settings.

Their new models (the 01-04 line) have [an even cooler way of adjusting the accuracy](http://www.technochas.ru/blog/wp-content/uploads/2017/10/WW-01-1.pdf). If you reset the seconds (just like you do in any Casio or in Sensor Watch) two times in an interval of 4h and 720h, the watch measures the deviation by itself, calculates by how much it should slow down/speed up, and automatically corrects itself for future reference. These models can also be manually adjusted in steps of 1/20th of a second per day.

So, my questions are a) if any of these options is viable, b) how hard could it be to implement them, and c) would anyone be willing to work on it since I do not have a clue where to start (beyond learning C from scratch)?

Thanks for your patience!

**Comment by tahnok** [comment]
I like this idea a lot! There's definitely room to improve on the accuracy at the moment, I've noticed my watch drifting. The idea of having the watch calculate it's own offset is really fascinating...

**Comment by joeycastillo** [comment]
I've thought about this a lot, and it's something I want to do; unfortunately, I don't have the equipment for it right now. My instinct though is that it would involve the temperature sensor board (included with campaign orders), a high resolution frequency counter, and either a GPS-disciplined TCXO or an atomic clock. The setup: 

* Solder two very fine wires to the sensor board's GND and A1 test points, and thread them outside the watch.
* Set up the frequency counter to measure the frequency of A1, with the external clock as a reference.
* [Enable the supply controller's OUT[0] output and enable the RTC Event Toggle](https://twitter.com/josecastillo/status/1503070912593149963). (copy/pastable code in alt text)
* Enable the RTC event output for the PEREO0 prescaler output, which generates an event 128 times a second.

At the end of this, A1 should be emitting a square wave with a frequency of about 64 Hz. Of course, you might see something like 64.000012 or 63.999987 Hz, which would cause the watch to drift. You'd want to build a watch face that shows the temperature and allows you to adjust the frequency correction register, then take notes on which FREQCORR values get the watch as close to 64.000000 Hz as possible. 

From there (and with some help from a refrigerator and a heater) you could create a series of FREQCORR values from, say, -20° C to 50° C, and code that lookup table into the watch. Then, out in the field, we'd be able to measure the temperature with the thermistor and know what frequency correction value to apply at a given temperature. 

Here are some notes on the frequency correction register, from the [SAM L22 data sheet](https://ww1.microchip.com/downloads/en/DeviceDoc/SAM-L22-Family-Data-Sheet-DS60001465B.pdf): 

<img width="1326" alt="Screen Shot 2022-06-15 at 9 31 17 PM" src="https://user-images.githubusercontent.com/1765152/173978550-cc09d5f0-ed1e-411a-ab16-b3e72a796d08.png">

**Reply by pebblewidower** [reply]
Well, that's a lot of work. So it looks like the Elektronika 55 approach is the easiest to implement. This 0.95367ppm resolution corresponds to 0.09059 seconds per day, which is very close to what the Elektronikas do.

The general idea is to display to the user a menu settable in steps of 0.1s (to round up and make the math easier for the user) and the firmware would interpret it as a FREQCORR.VALUE of 1 for every 0.1s added or subtracted, plus a negative or positive FREQCORR.SIGN depending on whether you want the watch to speed up or slow down. 

The FREQCORR.SIGN, I believe, would be the inverse of what is shown to the user, since if you want the watch to slow down a second per day you would input -1.0 in the settings, but the watch would actually add (positive FREQCORR.SIGN) 10 FREQCORR.VALUE, is that right?
